### ISSUE TITLE
Feature importance in R

### ISSUE THREAD

**Issue by tqchen** [tqchen/xgboost#123]
Since the feature importance function issue was discussed in the cross validation thread #114 , and it appears to me that it is worth to open an independent thread for the merging of feature importance function. Let us move new discussion  and enhancement with respect to feature importance to this thread

@pommedeterresautee @hetong007

**Comment by tqchen** [comment]
I think this is now in great shape:)   Thanks!

The unity branch will be merged into master in the new future. But yes I think it is OK to bring the linear model dump to master.

**Comment by pommedeterresautee** [comment]
Just for aesthetic and making a little bit easier parsing, would it be possible to add coma in these place (text format tree dump):
```
booster[0]:
0:[f0<1.00001] yes=1,no=2,missing=2 gain=9.00675,cover=21
	1:[f3<62.5] yes=3,no=4,missing=3 gain=0.588164,cover=10.75
		3:leaf=-1.36842cover=8.5
```
Between missing and gain, and between leaf and cover.

I am trying to write a tree plot function and I am wondering what is the meaning of the metric in front of word leaf? (is it homogeneity in the branch? enthropy? Gini?...). I am wondering because I am not sure it makes sense to include leaf in the tree.
Btw I will use this project https://github.com/knsv/mermaid for that purpose.

**Comment by tqchen** [comment]
The master is now updated to add the comma you mentioned.

The thing before leaf and split condition is the tree node index.  So root node is 0, the yes branch goes to node 1 and yes branch of node 1 goes to node 3.

**Comment by pommedeterresautee** [comment]
The way to make a path in the tree is one of the very few things I had understood by myself :) What i am wondering is the metric included in the leaf lines. What does it represent? I understand it is the end of a branch, so I imagine it s a metric of the data included in this branch. But what is the real meaning. 
Moreover there is a link between the branch but is there a link between the trees (i have edited my previous question). 

Kind regards,
Michaël

**Comment by tqchen** [comment]
for leaf branch, the only meaningful metric is the cover. Cover is defined by the sum of second order gradient of training data classified to the leaf, if it is square loss, this simply corresponds to the number of instances in that branch (that was why it is named as cover).

In terms of prediction, there is no big difference between RF and boosted tree, they are all ensemble tree models. boosted tree only differs from RF in terms of how you trained it. So they should have same model representation

You may be interested in  checkout http://homes.cs.washington.edu/~tqchen/pdf/BoostedTree.pdf for concept separation of model, objective and training

**Comment by pommedeterresautee** [comment]
Thanks for all these information.
I have written the code to convert this
```
booster[0]:
0:[f0<1.00001] yes=1,no=2,missing=2 gain=9.00675,cover=21
	1:[f3<62.5] yes=3,no=4,missing=3 gain=0.588164,cover=10.75
		3:leaf=-1.36842cover=8.5
		4:[f3<65.5] yes=7,no=8,missing=7 gain=0.307692,cover=2.25
			7:leaf=-0cover=1
			8:leaf=-0.666667cover=1.25
	2:[f3<39] yes=5,no=6,missing=5 gain=3.19787,cover=10.25
		5:leaf=-0.909091cover=1.75
		6:[f3<61.5] yes=9,no=10,missing=9 gain=5.62929,cover=8.5
			9:[f3<51.5] yes=11,no=12,missing=11 gain=0.405797,cover=4.75
				11:leaf=0.222222cover=1.25
				12:leaf=1.11111cover=3.5
			10:[f2<1.00001] yes=13,no=14,missing=14 gain=0.361134,cover=3.75
				13:leaf=-0.8cover=1.5
				14:[f3<67.5] yes=15,no=16,missing=15 gain=1.42308,cover=2.25
					15:leaf=0.5cover=1
					16:leaf=-0.666667cover=1.25
```
To that:
![rplot01](https://cloud.githubusercontent.com/assets/1029874/5606935/c3763c8c-9446-11e4-897f-c7d65ece1cfe.png)

I still need to wait a package to be pushed to Cran before pushing the plot function to Xgboost master. Moreover I have few questions (not all important):
- are you Ok with aesthetics choices? (told you not all question important)
- why for one hot encoded variable the text format model dump indicates <1.0001 instead of < 1 ?
- is there a way to only select the most interesting trees if there are plenty of them? I first thought looking to the sum of the gain in the tree, but I changed my mind because each tree is trained on a subset of observation and on a subset of features (from what I understood) but it would be very good to not have to plot 1000 trees for a big model. However, as per my understanding of the algorithm, first trees make most of the work, and final one focus on the part not yet learned by the previous one, so it would make sense to add a parameter which limits the plot to the n first trees, is that correct?
- **do you see errors in the tree I have generated?** The tree has been built based on the demo code I've [written](https://github.com/tqchen/xgboost/blob/master/R-package/demo/create_sparse_matrix.R). Basically, numbers are ages of patient, green they are ok, and the feature 3 is if they get a placebo or not.

Of course, I still need to replace the feature value ID by real values and add the cover in the leaf.

Not related, it's fascinating looking at model plotted, so interesting. In the example above, feature importance showed that two features were important, age of patient (very important) and placebo (just important). Here we can see that the first one used is placebo and not age (which is used almost everywhere else). In some way as a binary value, it s not surprising that using plaebo first provides more gain than splitting the age but still, it gives lots of understanding of the model.

**Comment by tqchen** [comment]
If the feature map is provided during dump and it contains the type of feature, then the dumped result will not contain the split value for the categorical feature. See dump.nice.txt if you run demo/binary_classification

**Comment by tqchen** [comment]
Since you mention visualization of a tree. You might be interested in this

https://github.com/CSE512-14W/fp-tianyizh-tqchen-luheng

This was a course project I did that visualize and allows you to interact with tree growing process. If you have a linux machine, you can checkout the project and play with it. The current version of xgboost no longer support the protocol due to refactoring. So you want to change installation script to

```bash
wget https://github.com/tqchen/xgboost/archive/v0.21.tar.gz
tar xf xgboost-0.21.tar.gz
mv xgboost-0.21 xgboost
cd xgboost;make;cd ..
./startserver.sh
```

**Comment by pommedeterresautee** [comment]
AWESOME!

From my experience at work, data visualization have a big impact on top management and clients (meaning -> people with $$$ to allocate to projects). Much more impact than an improvement of 10% in the accuracy of a model. And if the growing is done in live, it would be even better. I will give a try to your algo tonight (at work only Windows...) to see what it looks like.

On R there is this server tool: http://shiny.rstudio.com/gallery/ . I have seen plenty of very good stuff done with it. All users say it is easy to use. In some way, if current implementation of the algorithm gives me a way to access to the trees during the growing (through R), I should be able to come with an implementation in Shiny. From what I understand from this (document)[file:///C:/Users/mbenesty/Downloads/paper-tianyizh-tqchen-luheng.pdf] it will be less than what you did, but it will be still be possible to improve in the future.

Do you think that it would be difficult to have function showing me the model dump during the growing? 

Kind regards,
Michaël

**Comment by tqchen** [comment]
The original tool was done using D3 for visualization and cgi script. I think lots of existing visualization framework was based on D3. What we did for multiple trees was to collapse trees when there are too many of them and allow user to click and see any specific trees

**Comment by pommedeterresautee** [comment]
@tqchen I think this feature is correctly implemented for now. If you think it is, don't hesitate to close this issue :)

FYI I will add an importance feature plot function to xgboost I am using at work in the coming weeks (when work will be a bit less intensive). It's just classified horizontal histograms (it reads very well). I have added clustering of features (based on Gain) for color of bars (so you can easily see groups of important features). If you have other ideas, don't hesitate to share.

**Comment by tqchen** [comment]
Thanks @pommedeterresautee  for the great job! This is a nice addition to the package
